Simplify diffing branch rebased with geth master

elizabethengelman opened the pull request:
Updating this to capture how this branch is currently being used:
- used in the beta environment
- deployed on geth-statediff2 node

------------------------------------

Opening this PR against `upstream-master` which is based on geth's `master`, so we can see what a potential PR upstream would look like.

Companion PR in VDB to account for the change in the diff Payload struct: https://github.com/makerdao/vulcanizedb/pull/70

elizabethengelman committed:
core, eth, main, state, tests: Return state changes from statedb.Commit

elizabethengelman committed:
core: Send StateChangeEvents over a feed after being commited

elizabethengelman committed:
statediff: Add a statediff service

elizabethengelman committed:
main, utils, eth: Allow StateDiff service to be started via CLI

rmulhol commented on cmd/geth/config.go:
```diff
@@ -155,6 +155,11 @@ func makeFullNode(ctx *cli.Context) *node.Node {
+	if ctx.GlobalBool(utils.StateDiffFlag.Name) {
+		cfg.Eth.StateDiff = true
+		utils.RegisterStateDiffService(stack, ctx)
+	}
```
How different would this look if we removed the need for a CLI flag? Wondering if we could handle it the same way they do for a logs subscription where the node is always processing diffs but only sends/filters them if it receives a subscription via the rpc

elizabethengelman replied:
I'll start porting this over to using a pattern more similar to how logs are being handled, and we can see what it looks like. 

The one thing that comes to mind with removing the CLI flag is what if a user doesn't want to process statediffs (for performance reasons, etc) - maybe we can include a flag and have the default be turned on, and allow a user to turn off processing diffs if the want? 🤔

rmulhol replied:
Makes sense! Couldn't tell whether diffs would always be processed as part of the sync here (and then filtering would only kick in if there was a subscription)

rmulhol commented on core/blockchain.go:
```diff
@@ -1360,10 +1361,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
-		return NonStatTy, err
+	root, stateChanges, commitErr := state.Commit(bc.chainConfig.IsEIP158(block.Number()))
+	log.Debug("Sending StateChangeEvent to the feed", "block number", block.Number(), "count", len(stateChanges))
+	bc.stateChangeEventFeed.Send(StateChangeEvent{block, stateChanges})
```
Noticing that the send for logs on line 1464 has a check for `len(logs) > 0` - maybe we want a similar guard here?

elizabethengelman replied:
Good call!

elizabethengelman replied:
I've been struggling to find a way to test this. When I'm creating test transactions, there doesn't seem to be a case when there are 0 StateChanges, because I think the miner will always have their account balance increase. So I'm not sure that a guard here will do anything. Does that make sense? I'd be happy to rethink this/add a new story too!

rmulhol replied:
That makes sense to me!

rmulhol commented on core/state/statedb.go:
```diff
@@ -802,25 +802,43 @@ func (s *StateDB) clearJournalAndRefund() {
 			}
+
+			// Add the origin storage to the modifiedAccounts map before it's committed (and flushed from in-memory)
+			modifiedAccount.Storage = dirtyStateObject.originStorage
```
what's the rationale for using `dirtyStateObject` instead of `obj`?

also, does `originStorage` mean we're getting the storage from before the block's transactions were applied? (sorry should probably go look at the data structure 😅)

elizabethengelman replied:
Good call, it looks like `dirtyStateObject` and `obj` are the same here. 

w/r/t using `originStorage`, the `stateObject` has `originStorage`, `pendingStorage` and `dirtyStorage`. I would have assumed that `dirtyStorage` was the one I wanted, but when testing it out, I didn't get many diffs when I used `dirtyStorage`. It seemed like `originStorage` was the one that gave me results that matched up with the original patch version. Which is a little confusing. I want to do a bit more digging before we open this upstream, but this is definitely a question I'd like to put on that PR when we open it.

rmulhol commented on statediff/api.go:
```diff
@@ -0,0 +1,91 @@
+			case err := <-rpcSub.Err():
+				if err != nil {
+					log.Error("State diff service rpcSub error: " + err.Error())
+					err = api.sds.Unsubscribe(rpcSub.ID)
```
Ignore if this breaks convention for the repo, but maybe we could use a new name for this var so we're not overwriting the `rpcSub.Err()`

elizabethengelman replied:
Good call!

rmulhol commented on core/blockchain.go:
```diff
@@ -1360,10 +1361,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
-		return NonStatTy, err
+	root, stateChanges, commitErr := state.Commit(bc.chainConfig.IsEIP158(block.Number()))
+	log.Debug("Sending StateChangeEvent to the feed", "block number", block.Number(), "count", len(stateChanges))
+	bc.stateChangeEventFeed.Send(StateChangeEvent{block, stateChanges})
```
Also, looks like the `logsFeed` has a send [in the `reorg` function](https://github.com/ethereum/go-ethereum/blob/a75c0610b7aad63de01cefeeaf3426b332e718eb/core/blockchain.go#L2087) - I wonder if we might want to be sending from there as well. Maybe worth kicking off to a separate story though 🤔

elizabethengelman replied:
Added a story to account for this: https://makerdao.atlassian.net/browse/VDB-1274

rmulhol commented on statediff/api.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// Stream is the public method to setup a subscription that fires off statediff service payloads as they are created
+func (api *PublicStateDiffAPI) Stream(ctx context.Context) (*rpc.Subscription, error) {
```
Seems like subscriptions are usually setup in [the `filters` package](https://github.com/ethereum/go-ethereum/blob/d90d1db609c8d77baa422d49bd371207c06b4711/eth/filters/filter_system.go#L190) - I wonder if we might be able to move this stuff over there if we're no longer developing `statediff` stuff as a separate service

elizabethengelman replied:
Thanks for linking that. I think I hadn't been able to fully grasp the idea of doing the diff subscription following the pattern of logs, while also trying to get things simplified. But now, looking at it closer, I think that makes sense to follow that pattern over using a new service!  Nice looking out, and sorry it took me so long get it! 😜

elizabethengelman replied:
For reference, a new story for handling this: https://makerdao.atlassian.net/browse/VDB-1269

rmulhol commented on statediff/doc.go:
```diff
@@ -0,0 +1,58 @@
+	}
+}
+*/
+package statediff
```
Wondering if we can delete or need to update this file, assuming we're not doing intermediate nodes/paths and proofs/etc

rmulhol commented on statediff/service.go:
```diff
@@ -0,0 +1,337 @@
+	return false
+}
+
+func (sds *Service) filterByWatchedAddresses(stateChanges state.StateChanges) state.StateChanges {
```
normally I'm all in favor of moving this kind of logic to a separate package (as you've done here), but it seems like this kind of work is done (for logs) in the [`filters` package](https://github.com/ethereum/go-ethereum/blob/d90d1db609c8d77baa422d49bd371207c06b4711/eth/filters/filter_system.go#L311) - thinking maybe we want to move this there, but also wondering ways we could minimize the footprint. Maybe it would make sense to introduce a new go file in that package for stuff like this and `buildAccountDiff` 🤔

rmulhol commented on statediff/service.go:
```diff
@@ -0,0 +1,337 @@
+		BlockHash:   block.Hash(),
+	}
+
+	return rlp.EncodeToBytes(stateDiffWithoutUpdatedAccounts)
```
Probably nbd but I'm wondering whether we might be able to avoid doing this encoding for every block. Do you think we might be able to recognize a payload as empty earlier? Thinking for example that maybe `processStateChanges` could recognize if `StateDiff. UpdatedAccounts` is empty, and return a specific error/boolean so that we could assign `isEmpty` without needing to call `isEmptyPayload`

elizabethengelman replied:
👍 good eye!

rmulhol reviewed (commented):
Looks awesome! A lot of my comments are guided by comparing this against the logs subscription, but idk if that might be misguided in some places - excited to hear your thoughts!

karalabe committed:
eth: fix shutdown regression to abort downloads, not just cancel

karalabe committed:
Merge pull request #20988 from karalabe/catchup-shutdown

eth: fix shutdown regression to abort downloads, not just cancel

seh committed:
node: shut down all node-related HTTP servers gracefully (#20956)

Rather than just closing the underlying network listener to stop our
HTTP servers, use the graceful shutdown procedure, waiting for any
in-process requests to finish.

MariusVanDerWijden committed:
accounts/abi: added abi test cases, minor bug fixes  (#20903)

* accounts/abi: added documentation

* accounts/abi: reduced usage of arguments.LengthNonIndexed

* accounts/abi: simplified reflection logic

* accounts/abi: moved testjson data into global declaration

* accounts/abi: removed duplicate test cases

* accounts/abi: reworked abi tests

* accounts/abi: added more tests for abi packing

* accounts/abi/bind: refactored base tests

* accounts/abi: run pack tests as subtests

* accounts/abi: removed duplicate tests

* accounts/abi: removed unnused arguments.LengthNonIndexed

Due to refactors to the code, we do not need the arguments.LengthNonIndexed function anymore.
You can still get the length by calling len(arguments.NonIndexed())

* accounts/abi: added type test

* accounts/abi: modified unpack test to pack test

* accounts/abi: length check on arrayTy

* accounts/abi: test invalid abi

* accounts/abi: fixed rebase error

* accounts/abi: fixed rebase errors

* accounts/abi: removed unused definition

* accounts/abi: merged packing/unpacking tests

* accounts/abi: fixed [][][32]bytes encoding

* accounts/abi: added tuple test cases

* accounts/abi: renamed getMockLog -> newMockLog

* accounts/abi: removed duplicate test

* accounts/abi: bools -> booleans

BurtonQin committed:
core: add check in AddChildIndexer to avoid double lock (#20982)

This fixes a theoretical double lock condition which could occur in

    indexer.AddChildIndexer(indexer)

Nobody would ever do that though.

Co-authored-by: Felix Lange <fjl@twurst.com>

jyap808 committed:
rpc: make ExampleClientSubscription work with the geth API (#19483)

This corrects the call to eth_getBlockByNumber, which previously
returned this error:

  can't get latest block: missing value for required argument 1

Co-authored-by: Felix Lange <fjl@twurst.com>

fjl committed:
core: improve TestLogRebirth (#20961)

This is a resubmit of #20668 which rewrites the problematic test
without any additional goroutines. It also documents the test better.

The purpose of this test is checking whether log events are sent
correctly when importing blocks. The test was written at a time when
blockchain events were delivered asynchronously, making the check hard
to pull off. Now that core.BlockChain delivers events synchronously
during the call to InsertChain, the test can be simplified.

Co-authored-by: BurtonQin <bobbqqin@gmail.com>

ucwong committed:
go.mod: upgrade to golang-lru v0.5.4 (#20992)

golang-lru is now a go module, and the upgrade corrects a couple
of minor issues. In particular, the library could crash if you inserted
nil into an LRU cache.

seh committed:
rpc: add explicit 200 response for empty HTTP GET (#20952)

ucwong committed:
event, whisper/whisperv6: use defer where possible (#20940)

holiman committed:
signer, log: properly escape character sequences (#20987)

* signer: properly handle terminal escape characters

* log: use strconv conversion instead of custom escape function

* log: remove relection tests for nil

ucwong committed:
go.mod : goupnp v1.0.0 upgrade (#20996)

rjl493456442 committed:
core/state/snapshot: implement storage iterator (#20971)

* core/state/snapshot: implement storage iterator

* core/state/snapshot, tests: implement helper function

* core/state/snapshot: fix storage issue

If an account is deleted in the tx_1 but recreated in the tx_2,
the it can happen that in this diff layer, both destructedSet
and storageData records this account. In this case, the storage
iterator should be able to iterate the slots belong to new account
but disable further iteration in deeper layers(belong to old account)

* core/state/snapshot: address peter and martin's comment

* core/state: address comments

* core/state/snapshot: fix test

karalabe committed:
core/state/snapshot: fix journal nil deserialziation

karalabe committed:
Merge pull request #21003 from karalabe/snapshot-journal-nilfix

core/state/snapshot: fix journal nil deserialziation

rjl493456442 committed:
core/state/snapshot: fix trie generator reporter (#21004)

holiman committed:
accounts/external: fill account-cache if that hasn't already been done, fixes #20995 (#20998)

fjl committed:
go.sum: run go mod tidy (#21014)

BurtonQin committed:
les: remove invalid use of t.Fatal in TestHandshake (#21012)

holiman committed:
accounts/external: convert signature v value to 0/1 (#20997)

This fixes an issue with clef, which already transforms the signature
to use the legacy 27/28 encoding.

Fixes #20994

karalabe committed:
tests: cleanup snapshot generator goroutine leak

MariusVanDerWijden committed:
accounts/abi: removed Kind from Type struct (#21009)

* accounts/abi: removed Kind from Type struct

* accounts/abi: removed unused code

MariusVanDerWijden committed:
accounts/abi: move U256Bytes to common/math (#21020)

karalabe committed:
Merge pull request #21021 from karalabe/tests-snapshot-gen-cleanup

tests: cleanup snapshot generator goroutine leak

karalabe committed:
core/state/snapshot: release iterator after verification

karalabe committed:
Merge pull request #21023 from karalabe/snapshot-verify-iterator-release

core/state/snapshot: release iterator after verification

renaynay committed:
cmd/utils: renames flags related to http-rpc server (#20935)

* rpc flags related to starting http server renamed to http

* old rpc flags aliased and still functional

* pprof flags fixed

* renames gpo related flags

* linted

* renamed rpc flags for consistency and clarity

* added warn logs

* added more warn logs for all deprecated flags for consistency

* moves legacy flags to separate file, hides older flags under show-deprecated-flags command

* legacy prefix and moved some more legacy flags to legacy file

* fixed circular import

* added docs

* fixed imports lint error

* added notes about when flags were deprecated

* cmd/utils: group flags by deprecation date + reorder by date,

* modified deprecated comments for consistency, added warn log for --rpc

* making sure deprecated flags are still functional

* show-deprecated-flags command cleaned up

* fixed lint errors

* corrected merge conflict

* IsSet --> GlobalIsSet

* uncategorized flags, if not deprecated, displayed under misc

Co-authored-by: Martin Holst Swende <martin@swende.se>

karalabe committed:
tests: skip consensus test using 1GB RAM

karalabe committed:
cmd/geth: handle memfixes on 32bit arch with large RAM

karalabe committed:
build: raise test timeout back to 10 mins (#21027)

karalabe committed:
Merge pull request #21026 from karalabe/disable-highmem-test

tests: skip consensus test using 1GB RAM

MariusVanDerWijden committed:
account/abi: remove superfluous type checking (#21022)

* accounts/abi: added getType func to Type struct

* accounts/abi: fixed tuple unpack

* accounts/abi: removed type.Type

* accounts/abi: added comment

* accounts/abi: removed unused types

* accounts/abi: removed superfluous declarations

* accounts/abi: typo

karalabe committed:
eth: skip transaction announcer goroutine on eth<65

karalabe committed:
Merge pull request #21028 from karalabe/memfix-32bit-arch

cmd/geth: handle memfixes on 32bit arch with large RAM

karalabe committed:
Merge pull request #21032 from karalabe/skip-announce-goroutine-eth64

eth: skip transaction announcer goroutine on eth<65

ploui committed:
eth: don't inadvertently enable snapshots in archive nodes (#21025)

* eth: don't reassign more cache than is being made available

* eth: don't inadvertently enable snapshot in a case where --snapshot wasn't given

rjl493456442 committed:
trie: fix TestBadRangeProof unit test (#21034)

gballet committed:
eth/downloader: minor typo fixes in comments (#21035)

BurtonQin committed:
core: avoid double-lock in tx_pool_test (#20984)

karalabe committed:
core/state/snapshot: don't create storage list for non-existing accounts

karalabe committed:
Merge pull request #21036 from karalabe/snapshot-storage-leak

core/state/snapshot: don't create storage list for non-existing accounts

rjl493456442 committed:
core/state/snapshot: fix typo (#21037)

MariusVanDerWijden committed:
accounts/abi/bind: add void if no return args specified (#21002)

* accounts/abi/bind: add void if no return args specified

Currently the java generator generates invalid input on pure/view functions
that have no return type. e.g. `function f(uint u) view public {}`
This is not a problem in practice as people rarely ever write functions like this.

* accounts/abi/bind: use elseif instead of nested if

holiman committed:
core/state: include zero-address in state dump if present (#21038)

* Include 0x0000 address into the dump if it is present

* core/state: go fmt

Co-authored-by: Alexey Akhunov <akhounov@gmail.com>

holiman committed:
core/state: abort commit if read errors have occurred (#21039)

This finally adds the error check that the documentation of StateDB.dbErr
promises to do. dbErr was added in 9e5f03b6c (June 2017), and the check was
already missing in that commit. We somehow survived without it for three years.

elizabethengelman committed:
Update statedb test to return a storage value of 0

elizabethengelman committed:
Re-add fields to statediff types

elizabethengelman committed:
Fix linter and formatting issues

holiman committed:
core/state: avoid statedb.dbErr due to emptyCode (#21051)

* core/state: more verbose stateb errors

* core/state: fix flaw

* core/state: fixed lint

Co-authored-by: Marius van der Wijden <m.vanderwijden@live.de>

ligi committed:
build: fix CLI params for windows LNK files (#21055)

* build: Fix CLI params for windows LNK files

closes #21054

* Remove parameters

karalabe committed:
core/state: make GetCodeSize mirror GetCode implementation wise

rjl493456442 committed:
accounts/abi/bind/backend, internal/ethapi: recap gas limit with balance (#21043)

* accounts/abi/bind/backend, internal/ethapi: recap gas limit with balance

* accounts, internal: address comment and fix lint

* accounts, internal: extend log message

* tiny nits to format hexutil.Big and nil properly

Co-authored-by: Péter Szilágyi <peterke@gmail.com>

karalabe committed:
Merge pull request #21056 from karalabe/statedb-simpler-code

core/state: make GetCodeSize mirror GetCode implementation wise

MariusVanDerWijden committed:
params, cmd/utils: remove outdated discv5 bootnodes, deprecate flags (#20949)

* params: remove outdated discv5 bootnodes

* cmd/utils: deprecated bootnodesv4/v5 flags

holiman committed:
console: fix some crashes/errors in the bridge (#21050)

Fixes #21046

AusIV committed:
core/rawdb: stop freezer process as part of freezer.Close() (#21010)

* core/rawdb: Stop freezer process as part of freezer.Close()

When you call db.Close(), it was closing the leveldb database first,
then closing the freezer, but never stopping the freezer process.
This could cause the freezer to attempt to write to leveldb after
leveldb had been closed, leading to a crash with a non-zero exit code.

This change adds a quit channel to the freezer, and freezer.Close()
will not return until the freezer process has stopped.

Additionally, when you call freezerdb.Close(), it will close the
AncientStore before closing leveldb, to ensure that the freezer goroutine
will be stopped before leveldb is closed.

* core/rawdb: Fix formatting for golint

* core/rawdb: Use backoff flag to avoid repeating select

* core/rawdb: Include accidentally omitted backoff

fjl committed:
p2p: add 0 port check in dialer  (#21008)

* p2p: add low port check in dialer

We already have a check like this for UDP ports, add a similar one in
the dialer. This prevents dials to port zero and it's also an extra
layer of protection against spamming HTTP servers.

* p2p/discover: use errLowPort in v4 code

* p2p: change port check

* p2p: add comment

* p2p/simulations/adapters: ensure assigned port is in all node records

karalabe committed:
params: bump CHTs for the v1.9.14 release

holiman committed:
cmd, core, eth: background transaction indexing (#20302)

* cmd, core, eth: init tx lookup in background

* core/rawdb: tiny log fixes to make it clearer what's happening

* core, eth: fix rebase errors

* core/rawdb: make reindexing less generic, but more optimal

* rlp: implement rlp list iterator

* core/rawdb: new implementation of tx indexing/unindex using generic tx iterator and hashing rlp-data

* core/rawdb, cmd/utils: fix review concerns

* cmd/utils: fix merge issue

* core/rawdb: add some log formatting polishes

Co-authored-by: rjl493456442 <garyrong0905@gmail.com>
Co-authored-by: Péter Szilágyi <peterke@gmail.com>

karalabe committed:
Merge pull request #21061 from karalabe/cht-1.9.14

params: bump CHTs for the v1.9.14 release

holiman committed:
core: fixup blockchain tests (#21062)

core: fixup blockchain tests

rjl493456442 committed:
les: drop the message if the entire p2p connection is stuck (#21033)

* les: drop the message if the entire p2p connection is stuck

* les: fix lint

ucwong committed:
core/rawdb : log format fix for Unindexing transaction (#21064)

* core/rawdb : log format fix for Unindexing transaction

* core/rawdb: tiny fixup

Co-authored-by: Péter Szilágyi <peterke@gmail.com>

ucwong committed:
core/rawdb: remove unused math (#21065)

rjl493456442 committed:
miner: support disabling empty blockprecommits form the Go API (#20736)

* cmd, miner: add noempty-precommit flag

* cmd, miner: get rid of external flag

* miner: change bool to atomic int

* miner: fix tiny typo

Co-authored-by: Péter Szilágyi <peterke@gmail.com>

MariusVanDerWijden committed:
accounts/abi: accounts/abi/bind: Move topics to abi package (#21057)

* accounts/abi/bind: added test cases for waitDeployed

* accounts/abi/bind: added test case for boundContract

* accounts/abi/bind: removed unnecessary resolve methods

* accounts/abi: moved topics from /bind to /abi

* accounts/abi/bind: cleaned up format... functions

* accounts/abi: improved log message

* accounts/abi: added type tests

* accounts/abi/bind: remove superfluous template methods

MariusVanDerWijden committed:
accounts/abi: allow overloaded argument names (#21060)

* accounts/abi: allow overloaded argument names

In solidity it is possible to create the following contract:
```
contract Overloader {
    struct F { uint _f; uint __f; uint f; }
    function f(F memory f) public {}
}
```
This however resulted in a panic in the abi package.

* accounts/abi fixed error handling

rjl493456442 committed:
core: fix missing receipt on Clique crashes (#21045)

* core: fix missing receipt

* core: address comment

duanhao0814 committed:
ethstats: stop report ticker in each loop cycle #21070 (#21071)

Co-authored-by: Hao Duan <duan.hao@hyperchain.cn>

karalabe committed:
params: release go-ethereum v1.9.14

karalabe committed:
params: begin v1.9.15 release cycle

SatpalSandhu61 committed:
cmd: fix log contexts (#21077)

MariusVanDerWijden committed:
accounts/abi: simplify reflection logic (#21058)

* accounts/abi: simplified reflection logic

* accounts/abi: simplified reflection logic

* accounts/abi: removed unpack

* accounts/abi: removed comments

* accounts/abi: removed uneccessary complications

* accounts/abi: minor changes in error messages

* accounts/abi: removed unnused code

* accounts/abi: fixed indexed argument unpacking

* accounts/abi: removed superfluous test cases

This commit removes two test cases. The first one is trivially invalid as we have the same
test cases as passing in packing_test.go L375. The second one passes now,
because we don't need the mapArgNamesToStructFields in unpack_atomic anymore.
Checking for purely underscored arg names generally should not be something we do
as the abi/contract is generally out of the control of the user.

* accounts/abi: removed comments, debug println

* accounts/abi: added commented out code

* accounts/abi: addressed comments

* accounts/abi: remove unnecessary dst.CanSet check

* accounts/abi: added dst.CanSet checks

fjl committed:
console: fix unlockAccount argument count check (#21081)

holiman committed:
cmd/clef, signer/core: password input fixes (#20960)

* cmd/clef, signer/core: use better terminal input for passwords, make it possible to avoid boot-up warning

* all: move commonly used prompter to isolated (small) package

* cmd/clef: Add new --acceptWarn to clef README

* cmd/clef: rename flag 'acceptWarn' to 'suppress-bootwarn'

Co-authored-by: ligi <ligi@ligi.de>

BurtonQin committed:
accounts: add blockByNumberNoLock() to avoid double-lock (#20983)

* abi/bind/backends: testcase for double-lock

* accounts: add blockByNumberNoLock to avoid double-lock

* backend/simulated: use stateroot, not blockhash for retrieveing state

Co-authored-by: Martin Holst Swende <martin@swende.se>

ucwong committed:
whisper: use canonical import name of package go-ethereum (#21099)

rjl493456442 committed:
trie: extend range proofs with non-existence (#21000)

* trie: implement range proof with non-existent edge proof

* trie: fix cornercase

* trie: consider empty range

* trie: add singleSide test

* trie: support all-elements range proof

* trie: fix typo

* trie: tiny typos and formulations

Co-authored-by: Péter Szilágyi <peterke@gmail.com>
